src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaHelper.cs

```diff
@@ -0,0 +1,220 @@
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.Error(ex, "Error extracting propagated headers from Kafka message");
+                    }
+                }
+
+                string serviceName = tracer.Settings.GetServiceName(tracer, KafkaConstants.ServiceName);
+
+                var tags = new KafkaTags(SpanKinds.Consumer);
+
+                scope = tracer.StartActiveWithTags(KafkaConstants.ConsumeOperationName, parent: propagatedContext, tags: tags, serviceName: serviceName);
+
+                string resourceName = $"Consume Topic {(string.IsNullOrEmpty(topic) ? "kafka" : topic)}";
+
+                var span = scope.Span;
+                span.Type = SpanTypes.Queue;
+                span.ResourceName = resourceName;
+
+                if (partition is not null)
+                {
+                    tags.Partition = partition.ToString();
+                }
+
+                if (offset is not null)
+                {
+                    tags.Offset = offset.ToString();
+                }
+
+                if (message is not null && message.Timestamp.Type != TimestampType.NotAvailable)
+                {
+                    var consumeTime = span.StartTime.UtcDateTime;
+                    var produceTime = message.Timestamp.UtcDateTime;
+                    var queueTimeMs = Math.Max(0, (consumeTime - produceTime).TotalMilliseconds);
+                    // TODO: should this be a metric or tag?
+                    tags.RecordQueueTimeMs = queueTimeMs.ToString("0");
+                }
+
+                if (message is not null && message.Value is null)
+                {
+                    tags.Tombstone = "true";
+                }
+
+                // Consumer spans should always be measured
+                span.SetTag(Tags.Measured, "1");
+
+                tags.SetAnalyticsSampleRate(KafkaConstants.IntegrationId, tracer.Settings, enabledWithGlobalSetting: false);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            return scope;
+        }
+
+        internal static void CloseConsumerScope(Tracer tracer)
+        {
+            try
+            {
+                if (!tracer.Settings.IsIntegrationEnabled(KafkaConstants.IntegrationId)
+                    || !tracer.Settings.KafkaCreateConsumerScopeEnabled)
+                {
+                    // integration disabled, skip this trace
+                    return;
+                }
+
+                var activeScope = tracer.ActiveScope;
+                var currentSpan = activeScope?.Span;
+                if (currentSpan?.OperationName != KafkaConstants.ConsumeOperationName)
+                {
+                    // Not currently in a consumer operation
+                    return;
+                }
+
+                // TODO: record end-to-end time?
```

Comment:
The end-to-end time is already recorded as the Span duration. Did you have something else in mind?

Reply:
The Java implementation records the _trace_ end-to-end time. That requires passing the trace start time around as baggage (as the start will be a distributed trace!) Seems like a lot of work, and to my mind would require sending that data in _all_ distributed spans, which touches a lot of places. I'm not convinced of the utility, what do you think?

https://github.com/DataDog/dd-trace-java/blob/e5f2514a90b04f4f7620d1497207b5a0224d0a3e/dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/KafkaDecorator.java#L93-L105

Reply:
> the .NET Tracer doesn't have baggage support yet

Well that solves that then 😄